### PR TITLE
Adding validation for cluster owner and cluster member v1 endpoints

### PR DESCRIPTION
**Pull request by anupama2501**
## Problem
Currently we do not have test suite to validate the P0 use cases for cluster owners/ cluster members that touches v1 endpoints 

 
## Solution
In this PR we are creating a test suite that covers P0 test cases for RBAC cluster owner and cluster member that touches v1 endpoints. Following are the test scenarios covered in the PR:

### Automated Testing

**Cluster Owner:**

1. Standard user cannot list downstream cluster
2. Cluster owner can list the downstream cluster
3. Cluster owner should be able to list projects the user does not own
4. Cluster owner should be able to list namespaces the user does not own
5. Cluster owner should be able to create namespaces in Projects the user does not own
6. Cluster owner should be able to create Projects in the cluster
7. Cluster owner should be able to delete the namespaces they are not owner
8. Cluster owner should be able to delete the projects they are not owner

**Cluster Member:**

1. Cluster member owner can list the downstream cluster
2. Cluster member should NOT be able to list projects the user does not own 
3. Cluster member should NOT be able to list namespaces the user does not own 
4. Cluster member should NOT be able to create namespaces in Projects the user does not own
5. Cluster member should be able to create Projects in the cluster 
6. Cluster Member should be able to delete the namespaces they are not owner
7. Cluster Member should be able to delete the projects they are not owner

### CODE REVIEW

File: tests/framework/extensions/namespaces/namespaces.go
Comment (igomez06) [inline tests/framework/extensions/namespaces/namespaces.go:147]: Anu have this function convert the unstructured list to a namespace list. There are example in the other extensions how to do this. We don't want to return these functions to return unstructureds. But actually Ugur has a PR for this, that we'll get merged in first https://github.com/rancher/rancher/pull/38171

File: tests/framework/extensions/projects/projects.go
Comment (igomez06) [inline tests/framework/extensions/projects/projects.go:55]: Do you think we really need a helper function to delete a project? I don't think it's needed. a delete call is pretty simple to use

File: tests/v2/validation/rbac/rbac.go
Comment (igomez06) [inline tests/v2/validation/rbac/rbac.go:18]: We have this function somewhere else. I think it's provisioning there is no need to rewrite the function.

File: tests/v2/validation/rbac/rbac_test.go
Comment (igomez06) [inline tests/v2/validation/rbac/rbac_test.go:80]: Remove all the fmt.Println statements and use the logging instead. Either rb.T().Log or Logrus.Info(). No Printlns at all

File: tests/v2/validation/rbac/rbac_test.go
Comment (igomez06) [inline tests/v2/validation/rbac/rbac_test.go:162]: Fix the formatting on the if/else if statements. The blocks aren't indented and there is no space between the statement and the `{` it's hard to read

File: tests/v2/validation/rbac/rbac_test.go
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_test.go:120]: is there a 3rd or more cases we may add here that we should catch in an `else` statement? it looks strange to have `if` and `if else` but no `else`.

File: tests/v2/validation/rbac/rbac_test.go
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_test.go:189]: why 6 seconds? I think this should be using `wait.WatchWait` instead, if you're waiting on the user to remove?

File: tests/v2/validation/rbac/rbac_test.go
Comment (slickwarren) [inline tests/v2/validation/rbac/rbac_test.go:224]: not sure what the sleep is for

File: tests/framework/extensions/namespaces/namespaces.go
Comment (igomez06) [inline tests/framework/extensions/namespaces/namespaces.go:147]: Hey Anu Ugur has a PR with this helper, which I would like to be merged and used instead. 
https://github.com/rancher/rancher/pull/38171

File: tests/v2/validation/rbac/rbac_test.go
Comment (igomez06) [reply tests/v2/validation/rbac/rbac_test.go:189]: I agree use WatchWait there shouldn't be a reason to have a sleep

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go:120]: We only have 2 roles for cluster-owner and cluster-member so either if this role or that role.

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go:189]: Even after adding a WatchWait, test errors out as the user still has a list of clusters. Therefore added the time.Sleep. I'll look more into it probably using an incorrect resource wait.

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go:224]: Similar to the above adding a WatchWait, test errors out as the user is not updated with the permissions. Therefore added the time.Sleep. I'll look more into it probably using an incorrect resource wait.

File: tests/v2/validation/rbac/rbac.go
Comment (sowmyav27) [inline tests/v2/validation/rbac/rbac.go:15]: We should have some existing function for this isnt it since we have other tests which use standard user to execute? Why are we creating a new helper function? Can we use existing one?

File: tests/v2/validation/rbac/rbac.go
Comment (sowmyav27) [inline tests/v2/validation/rbac/rbac.go:36]: There is a `GetProjectList` vs a `GetListProjects`. What is the difference between both? Can we use one of them?

File: tests/v2/validation/rbac/rbac.go
Comment (sowmyav27) [inline tests/v2/validation/rbac/rbac.go:67]: Why `testproject-CO` ? Why `CO`? Can we add a generic one instead?

File: tests/v2/validation/rbac/rbac.go
Comment (sowmyav27) [inline tests/v2/validation/rbac/rbac.go:53]: Can we change this to `ListNamespace` instead? or `GetNamespace` ? Or is this how it's everywhere else in the framework ?

File: tests/v2/validation/rbac/rbac_test.go
Comment (sowmyav27) [inline tests/v2/validation/rbac/rbac_test.go:56]: Can this be extended to Project Owner, Project member and restricted admin user roles also ? 

Also this function name `ValidateClustersList` should be a more appropriate one, since this covers about 8 tests.

**Review by sowmyav27 — CHANGES_REQUESTED** [review]

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go:56]: `ValidateClustersList` does not cover 8 test cases. The purpose of this function is to validate the cluster list obtained as a cluster member/owners/standard users.

Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go:56]: Yes. This can be extended to Project Owner, project members and restricted admins as well.

File: tests/v2/validation/rbac/rbac.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac.go:53]: It is GetNamespacesList - in general in the framework. But I changed it to GetNamespaces to be more generic

File: tests/v2/validation/rbac/rbac.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac.go:36]: GetProjectList calls the project list function and the result is obtained as a list. In GetListProjects, I modify the list of projects obtained and return the names of the projects only as a slice.

File: tests/v2/validation/rbac/rbac_test.go
Comment (igomez06) [reply tests/v2/validation/rbac/rbac_test.go:56]: Quick question, this function is agnostic to the role of the user, correct?

File: tests/framework/extensions/namespaces/namespaces.go
Comment (caliskanugur) [reply tests/framework/extensions/namespaces/namespaces.go:49]: The last PR about this [#38171](https://github.com/rancher/rancher/pull/38171). Seems like there's not an implemented delete function, however, namespace creation function already has its session cleanup [here](https://github.com/rancher/rancher/blob/20a4ab0b4e3d19ce23b94227985f03c0ca05ecaf/tests/framework/extensions/namespaces/create.go#L92-L117). I am unsure if overwriting and requesting delete call twice would cause a problem for cleanup enabled test runs.

Comment (igomez06) [reply tests/framework/extensions/namespaces/namespaces.go:49]: That's fine, Ugur, the cleanup would just log an error or not, but the delete needs to be explicitly tested, so never mind, thanks for answering.

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go:90]: I modified this as per your suggestion. Initially I had planned the test cases to be all in one function as each test case depended on the previous one. I changed them to be smaller test case to be more readable.

File: tests/v2/validation/rbac/rbac.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac.go]: This needs to be in the rbac package. Changed the formatting to be clusterProject

File: tests/v2/validation/rbac/rbac.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac.go]: This needs to be in the rbac package. Changed the formatting to be getNamespaces

File: tests/v2/validation/rbac/rbac.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac.go]: This needs to be in the rbac package. Changed the formatting to be listProjects

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go]: yes, its agnostic to the role of the user

File: tests/v2/validation/rbac/rbac_test.go
Comment (caliskanugur) [inline tests/v2/validation/rbac/rbac_test.go:93]: I think logging and comments are really explanatory, there's only one thing I want to point out, but I don't believe that this is necessary. Go test has its own logging too, so in the tests, this could also be used `rb.T().Logf("Created project as a %v is %v", role, createProjectAsClusterMembers.Name)`

File: tests/v2/validation/rbac/rbac_test.go
Comment (caliskanugur) [inline tests/v2/validation/rbac/rbac_test.go:202]: Just out of curiosity, would a new sub-session be attached to the globally used client if the returned client isn't used for the upcoming client calls, I am unsure about that. What should be expected in this case @igomez06?

**Review by caliskanugur — COMMENTED** [review]
A few questions and nitpicks, but looks really good!

A small note: I believe there's an issue with your Go code formatter. So your IDE either needs the Go extension or if you have it, this might be the disabled auto-save format option.

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go]: I believe I do not have to re initialize the session. So, I removed the `_, err= rb.standardUserClient.WithSession(subSession)`

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go:90]: Sure. thanks Ugur! I haven't changed the current logging for the validation calls but I used it in the TestRbac call.

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go:21]: I'm not sure I understand why it has to be made public?

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go]: Good point. Changed it to rbac.go

File: tests/v2/validation/rbac/rbac.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac.go]: Thanks for pointing that out. Updated.

File: tests/v2/validation/rbac/rbac_test.go
Comment (igomez06) [reply tests/v2/validation/rbac/rbac_test.go:22]: It's standard Anu, the test structs should be public

File: tests/v2/validation/rbac/rbac_test.go
Comment (igomez06) [reply tests/v2/validation/rbac/rbac_test.go]: Perfect

File: tests/v2/validation/rbac/rbac.go
Comment (igomez06) [inline tests/v2/validation/rbac/rbac.go:61]: Just to point out, this will probably not be needed once the v1 Steve client is merged

File: tests/v2/validation/rbac/rbac_test.go
Comment (igomez06) [inline tests/v2/validation/rbac/rbac_test.go:100]: I agree with Ugur, having these log messages in conjunction with the name of the test is a little redunant

File: tests/v2/validation/rbac/rbac_test.go
Comment (igomez06) [inline tests/v2/validation/rbac/rbac_test.go:184]: We don't need this block of code to be within a Run(). But if you want to keep it, can you please make it more descriptive? In Jenkins it just looks like `Cluster_Owner#1` etc. May something like "`Set Up User With Cluster Role ..." `That way know what aspect of the test failed on first glance

File: tests/v2/validation/rbac/rbac_test.go
Comment (igomez06) [inline tests/v2/validation/rbac/rbac_test.go:201]: Update name remove log message

**Review by igomez06 — COMMENTED** [review]
Looks good some nitpicks. But would like for this PR:https://github.com/rancher/rancher/pull/39341
 to be merged in first before we get this in.

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go:100]: I have removed the log messages for validations that have a test name. but for this test, since its a more generic case `ValidateNS` which includes 3 test cases, I have kept the log messages for these 3 test cases.

File: tests/v2/validation/rbac/rbac_test.go
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_test.go:208]: @igomez06 Can we not wait for the PR https://github.com/rancher/rancher/pull/39341 to be merged in and merge this PR if everything looks ok ? Since we want a few test cases for RBAc to be addressed in v2.7.0 and not have them in this PR.

File: tests/v2/validation/rbac/rbac_test.go
Comment (igomez06) [inline tests/v2/validation/rbac/rbac_test.go:230]: Typo Test Case

File: tests/v2/validation/rbac/rbac_test.go
Comment (igomez06) [reply tests/v2/validation/rbac/rbac_test.go:208]: Well Anu, PR https://github.com/rancher/rancher/pull/39341 is vital for v1 calls, and delaying more not only adds more work for me, but it'll keep pushing this off, which is super critical to our testing. You can branch off of this branch and add tests.

**Comment by caliskanugur** [comment]
Anupama I think PR needs to rebase/synch, could you rebase the PR with the 2.6 branch

**Comment by anupama2501** [comment]
Tracked here https://github.com/rancher/rancher/pull/39630

**Pull request closed without merging**
